public/app/plugins/datasource/loki/addToQuery.ts

```diff
@@ -39,22 +39,36 @@ export function addLabelToQuery(query: string, key: string, operator: string, va
   }
 }
 
-type StreamSelectorPosition = { from: number; to: number; query: LokiVisualQuery };
-type PipelineStagePosition = { from: number; to: number };
+/**
+ * Adds parser to existing query. Useful for query modification for hints.
+ * It uses LogQL parser to find instances of stream selectors or line filters and adds parser after them.
+ *
+ * @param query
+ * @param parser
+ */
+export function addParserToQuery(query: string, parser: string): string {
+  const lineFilterPositions = getLineFiltersPositions(query);
+
+  if (lineFilterPositions.length) {
+    return addParser(query, lineFilterPositions, parser);
```

Comment:
If we have lineFilters, we add parser after them. Otherwise we add it right after stream selector.

Reply:
It looks like this is adding one parser for each 2 line-filters:


https://user-images.githubusercontent.com/8092184/177737578-e4830491-86c8-4ac6-9af4-7e7e79a82feb.mov

But this looks more like a bug in the the leezer `Parser` for me?

Reply:
It can be correct (and in this case it is), if you have mixed log lines - as in this case where we have some lines in `logfmt` and some in json `format`

Reply:
> I think we might miss a s here https://github.com/grafana/lezer-logql/blob/main/src/logql.grammar#L56 ? But I am unsure.

Yes, that fixes it. I will go ahead and add a PR for that - if that's not correct, we can just decline that. :)

Reply:
Awesome! Thanks for fixing this. Now when new version is published, you can just create a new PR to update this dependency in grafana repo.

Reply:
Upgrade PR is https://github.com/grafana/grafana/pull/51888.